src/boot/chid.h

```diff
@@ -2,22 +2,56 @@
 #pragma once
 
 #include "efi.h"
-
 #include "chid-fundamental.h"
 
+/* A .hwids PE section consists of a series of 'Device' structures. A 'Device' structure binds a CHID to some
+ * resource, for now only Devicetree blobs. Designed to be extensible to other types of resources, should the
+ * need arise. The series of 'Device' structures is followed by some space for strings that can be referenced
+ * by offset by the Device structures. */
+
+enum {
+        DEVICE_TYPE_DEVICETREE = 0x1, /* A devicetree blob */
+
+        /* Maybe later additional types for:
+         *   - CoCo Bring-Your-Own-Firmware
+         *   - ACPI DSDT Overrides
+         *   - … */
+};
+
+#define DEVICE_SIZE_FROM_DESCRIPTOR(u) ((uint32_t) (u) & UINT32_C(0x0FFFFFFF))
+#define DEVICE_TYPE_FROM_DESCRIPTOR(u) ((uint32_t) (u) >> 28)
+#define DEVICE_MAKE_DESCRIPTOR(type, size) (((uint32_t) (size) | ((uint32_t) type << 28)))
+
+#define DEVICE_DESCRIPTOR_DEVICETREE DEVICE_MAKE_DESCRIPTOR(DEVICE_TYPE_DEVICETREE, sizeof(Device))
+#define DEVICE_DESCRIPTOR_EOL UINT32_C(0)
+
 typedef struct Device {
-        uint32_t struct_size;       /* = sizeof(struct Device), or 0 for EOL */
-        uint32_t name_offset;       /* nul-terminated string or 0 if not present */
-        uint32_t compatible_offset; /* nul-terminated string or 0 if not present */
+        uint32_t descriptor; /* The highest four bit encode the type of entry, the other 28 bit encode the
+                              * size of the structure. Use the macros above to generate or take apart this
+                              * field. */
         EFI_GUID chid;
+        union {
+                struct {
+                        /* These offsets are relative to the beginning of the .hwids PE section. */
+                        uint32_t name_offset;          /* nul-terminated string or 0 if not present */
+                        uint32_t compatible_offset;    /* nul-terminated string or 0 if not present */
+                } devicetree;
+                /* fields for other types… */
+        };
 } _packed_ Device;
 
```

Comment:
Should there be an assert for size here and a comment that this definition is used in ukify and needs to be kept in sync? 
```
assert_cc(sizeof(Device) == 28);
```
This is useful as a documentation and a sanity check if other types are to be added.

Reply:
It [seems](https://stackoverflow.com/a/71425195/13293614) that `_packed_` works a bit strange with unions, so maybe also add checks for field offsets. These are basically hardcoded in ukify as of now.

Reply:
uh, i am only realizing now that the ukify part has been merged by now. i guess that needs updating too, given that the order of the fields changed.

@anonymix007 any chance you can fix the ukify part to match the new field order and value? I cannot really test this here. 

When I prepped the PR I was assuming your ukify patch hadn't been merged yet, so I wanted to dump that problem to catch up with this on you anyway. That didn#t work really though ;-)

Reply:
Sure, will do.

Reply:
@anonymix007 please open another PR

Reply:
@anonymix007 will you be able to open a PR today? It's the last thing blocking RC2 - thanks

Reply:
@bluca #35179
Note that I haven't actually boot tested it, only did a simple userspace test.